dfsclient/src/main/java/org/hpccsystems/dfs/client/DataPartition.java

```diff
@@ -266,6 +269,38 @@ public String getCopyIP(int copyindex)
         return copyLocations[copyindex];
     }
 
+    /**
+     * Set the copy IP
+     *
+     * @param copyIndex
+     *            the copyindex
+     * @param copyIP The IP of the file part copy
+     */
+    public void setCopyIP(int copyIndex, String copyIP)
+    {
+        if (copyIndex < 0 || copyIndex >= copyLocations.length) return;
+
+        copyLocations[copyIndex] = copyIP;
+    }
+
+    /**
+     * Add file part copy
+     *
+     * @param index The index at which to insert the file part copy
+     * @param copyIP The IP of the new file part copy
+     * @param copyPath The path of the new file part copy
+     */
+    public void add(int index, String copyIP, String copyPath)
+    {
+        List<String> copyLocationsList = new ArrayList<>(Arrays.asList(copyLocations));
+        copyLocationsList.add(index, copyIP);
+        copyLocations = copyLocationsList.toArray(new String[0]);
```

Comment:
Agreed, the existing type is a simple array; I had considered moving this to a List but other parts of the code are expecting an array, so it would require the internal list be converted to an Array at those locations. I thought it would be better to do the conversion here in the add() function because this unlikely to be used outside of test cases.